crypto/s2n_certificate.c

```diff
@@ -605,3 +611,113 @@ int s2n_get_cert_der(const struct s2n_cert *cert, const uint8_t **out_cert_der,
+    POSIX_ENSURE_REF(internal_data);
+    POSIX_CHECKED_MEMCPY(*out_data, internal_data, (*out_len));
+
+    return S2N_SUCCESS;
+}
+
+int s2n_get_x509_extension_oid_value(struct s2n_cert *cert, const uint8_t *oid_field_in, const uint32_t oid_field_in_len,
+                                      uint8_t **oid_value_out, uint32_t *oid_value_out_len, bool *critical)
+{
+    POSIX_ENSURE_REF(cert);
+    POSIX_ENSURE_REF(oid_field_in);
+    POSIX_ENSURE_GT(oid_field_in_len, 0);
+    POSIX_ENSURE_REF(oid_value_out);
+    POSIX_ENSURE_REF(oid_value_out_len);
+    POSIX_ENSURE_REF(critical);
+    POSIX_ENSURE_REF(cert->raw.data);
+
+    uint8_t *der_in = cert->raw.data;
+    /* Obtain the openssl X509 cert from the ASN1 DER certificate input. 
+     * Note that d2i_X509 increments *der_in to the byte following the parsed data.
+     * Ref: https://www.openssl.org/docs/man1.1.0/man3/d2i_X509.html.
+     */
+    DEFER_CLEANUP(X509 *x509_cert = d2i_X509(NULL, (const unsigned char **)(void *)&der_in, cert->raw.size),
+                  X509_free_pointer);
+    POSIX_ENSURE_REF(x509_cert);
+
+    /* Retrieve the number of x509 extensions present in the certificate */
+    int ext_count = X509_get_ext_count(x509_cert);
+    POSIX_ENSURE_GT(ext_count, 0);
+
+    for (size_t loc = 0; loc < ext_count; loc++) {
+        ASN1_OCTET_STRING *oid_asn1_str = NULL;
+        bool match_found = false; 
+
+        /* Retrieve the x509 extension at location loc */
+        X509_EXTENSION *x509_ext = X509_get_ext(x509_cert, loc);
+        POSIX_ENSURE_REF(x509_ext);
+
+        /* Retrieve the x509 extension's critical value */
+        *critical = X509_EXTENSION_get_critical(x509_ext);
+
+        /* Retrieve the extension object/OID/extnId */
+        ASN1_OBJECT *asn1_obj = X509_EXTENSION_get_object(x509_ext);
+        POSIX_ENSURE_REF(asn1_obj);
+        int nid_from_cert = OBJ_obj2nid(asn1_obj);
+        if (nid_from_cert == NID_undef) {
+            char oid_field[OID_FIELD_MAX_LEN] = { 0 };
+            /* 0 here refers to preference of textual representation over numerical one */
+            int oid_len = OBJ_obj2txt(oid_field, sizeof(oid_field), asn1_obj, 0);
+            match_found = (oid_len == oid_field_in_len) && (strcmp(oid_field, (const char *)oid_field_in) == 0);
+        } else { 
+            int nid_from_in = OBJ_txt2nid((const char *)oid_field_in);
+            match_found = (nid_from_in == nid_from_cert); 
+        }
+        /* If match found, retrieve the corresponding OID value for the x509 extension */
+        if (match_found) {
+            /* X509_EXTENSION_get_data() returns the data of extension ex. 
+             * The returned pointer is an internal value which must not be freed up. 
+             */
+            oid_asn1_str = X509_EXTENSION_get_data(x509_ext);
+            POSIX_ENSURE_REF(oid_asn1_str);
+            *oid_value_out_len = ASN1_STRING_length(oid_asn1_str);
+            POSIX_ENSURE_GT(*oid_value_out_len, 0);
```

Comment:
You need to make sure the provided buffer has enough space before setting the length. Otherwise you're going to get a buffer overwrite